dutch-auction/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+#![no_std]
+#![feature(const_btree_new)]
+
+use codec::Encode;
+use gstd::{msg, prelude::*, exec::block_timestamp, ActorId};
+use primitive_types::U256;
+use nft_example_io;
+
+pub mod state;
+pub use state::{State, StateReply};
+
+pub use auction_io::{Action, Event, InitConfig, CreateConfig};
+
+const ZERO_ID: ActorId = ActorId::new([0u8; 32]);
+const DURATION: u64 = 7 * 24 * 60 * 60 * 1000;
+
+#[derive(Debug)]
+pub struct NFT {
+    pub token_id: U256,
+    pub owner: ActorId,
+    pub contract_id: ActorId,
+}
+
+#[derive(Debug)]
+pub struct Auction {
+    pub nft: NFT,
+    pub starting_price: U256,
+    pub discount_rate: U256,
+    pub is_using: bool,
+    pub start_at: u64,
+    pub expires_at: u64,
+}
+
+static mut CONTRACT: Auction = Auction {
+    nft: NFT {
+        token_id: U256::zero(),
+        owner: ZERO_ID,
+        contract_id: ZERO_ID,
+    },
+    starting_price: U256::zero(),
+    discount_rate: U256::zero(),
+    is_using: false,
+    start_at: 0,
+    expires_at: 0,
+};
+
+impl Auction {
+    fn buy(&mut self) {
+        if !self.is_using {
+            panic!("already bought or auction expired");
+        }
+
+        if block_timestamp() >= self.expires_at {
+            panic!("auction expired");
+        }
+
+        let price = self.token_price().as_u128();
+
+        if msg::value() < price {
+            panic!("value < price");
+        }
+
+        self.is_using = false;
+        let refund = msg::value() - price;
+
+        msg::send(
```

Comment:
It's better to use asynchronous function `send_and_wait_for_reply` when you are interaction with another program